ow_plexil/src/plexil-adapter/FaultDependencies.cpp

```diff
@@ -0,0 +1,355 @@
+
+  // If System is specified we return all active faults across every subsystem
+  if (name == "System"){
+    for (auto i: m_faults){
+      if (i.second.faulty == 1 && active_faults.size() <= 10){
+        active_faults.push_back(i.first);
+      }
+    }
+  }
+  // If no subsystem by name exists
+  else if(m_subsystems.find(name) == m_subsystems.end()){
+    ROS_WARN("SUBSYSTEM: %s DOES NOT EXIST.", name.c_str());
+  }
+  // Get all active faults in specified subsystem
+  else{
+    for (auto i: m_subsystems[name].faults){
+      if (m_faults[i].faulty == 1 && active_faults.size() <= 10){
+        active_faults.push_back(i);
+      }
+    }
+  }
+  return active_faults;
+}
+
+void FaultDependencies::updateFault(const std::string &name, int status){
+  // If a fault doesnt exist
+  if (m_faults.find(name) == m_faults.end()){
+    return;
+  }
+  // If a fault exists but does not need to be updated
+  if (m_faults[name].faulty == status){
+    return;
+  }
+
+  m_faults[name].faulty = status;
+  std::string parent_subsystem  = m_faults[name].parent_subsystem;
+
+
+  // increment local fault flag for parent subsystem and change faulty flag as needed
+  if (status == 1){
+    m_subsystems[parent_subsystem].local_fault += 1;
+  }
+  else{
+    m_subsystems[parent_subsystem].local_fault -= 1; 
+  }
+
+  if (m_subsystems[parent_subsystem].local_fault > 0){
+    m_subsystems[parent_subsystem].faulty = 1; 
+  }
+  else{
+    m_subsystems[parent_subsystem].faulty = 0; 
+  }
+
+  for (auto dep: m_faults[name].impacts){
+    // If an impact to update is a subsystem
+    if (dep.second == "Subsystem"){
+      if (m_subsystems.find(dep.first) != m_subsystems.end()){
+        updateSubsystem(dep.first, status, parent_subsystem);
+      }
+    }
+    // Otherwise we treat it as a procedure to update
+    else{
+      if (m_procedures.find(dep.first) != m_procedures.end()){
+        if (status == 1){
+            m_procedures[dep.first].inoperable += 1;
+        }
+        else{
+            m_procedures[dep.first].inoperable -= 1;
+        }
+      }
+    }
+ 
+  }
+  // If verbose output flag is true we print the current fault dependency info
+  if (m_verbose_flag){
+    DebugPrint();
+  }
+}
+
+void FaultDependencies::updateSubsystem(const std::string &name, int status, const std::string &parent){
+  int prev_inoperable_state = m_subsystems[name].inoperable;
+  // increment non_local fault and inoperable flags if status is 1
+  if (status == 1){
+      if (parent != name){
+        m_subsystems[name].non_local_fault += 1;
+      }
+      m_subsystems[name].inoperable += 1;
+      ROS_INFO("SUBSYSTEM: %s INOPERABLE.", name.c_str());
+  }
+  // otherwise decrement
+  else{
+      if (parent != name){
+        m_subsystems[name].non_local_fault -= 1;
+      }
+      m_subsystems[name].inoperable -= 1;
+      if (m_subsystems[name].inoperable == 0){
+        ROS_INFO("SUBSYSTEM: %s NOW OPERABLE.", name.c_str());
+      }
+  }
```

Comment:
This same pattern is used in `cascadeFaults`. 
```suggestion
  m_subsystems[name].update(status);
```